pkg/operator/agent/agent.go

```diff
@@ -0,0 +1,250 @@
+						},
+						{
+							Name: "libmodules",
+							VolumeSource: v1.VolumeSource{
+								HostPath: &v1.HostPathVolumeSource{
+									Path: "/lib/modules",
+								},
+							},
+						},
+					},
+					HostNetwork: true,
+					Tolerations: []v1.Toleration{
+						{
+							Effect:   v1.TaintEffectNoSchedule,
+							Operator: v1.TolerationOpExists,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	_, err = a.clientset.Extensions().DaemonSets(namespace).Create(ds)
+	if err != nil {
+		if !kserrors.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create rook-agent daemon set. %+v", err)
+		}
+		logger.Infof("rook-agent daemonset already exists")
+	} else {
+		logger.Infof("rook-agent daemonset started")
+	}
+	return nil
+
+}
+
+func (a *Agent) discoverFlexvolumeDir() string {
+	//copy flexvolume to flexvolume dir
+	nodeName := os.Getenv(k8sutil.NodeNameEnvVar)
+	if nodeName == "" {
+		logger.Warningf("cannot detect the node name. Please provide using the downward API in the rook operator manifest file")
+		return getDefaultFlexvolumeDir()
+	}
+
+	// determining where the path of the flexvolume dir on the node
+	var flexvolumeDirPath string
+	nodeConfigURI, err := k8sutil.NodeConfigURI()
+	if err != nil {
+		logger.Warningf(err.Error())
+		return getDefaultFlexvolumeDir()
+	}
+	nodeConfig, err := a.clientset.Core().RESTClient().Get().RequestURI(nodeConfigURI).DoRaw()
+	if err != nil {
+		logger.Warningf("unable to query node configuration: %v", err)
+		return getDefaultFlexvolumeDir()
+	}
+
+	// unmarshall to a NodeConfigKubelet
+	configKubelet := NodeConfigKubelet{}
+	if err := json.Unmarshal(nodeConfig, &configKubelet); err != nil {
+		logger.Warningf("unable to parse node config from Kubelet: %+v", err)
+		return getDefaultFlexvolumeDir()
+	}
+	flexvolumeDirPath = configKubelet.ComponentConfig.VolumePluginDir
+	if flexvolumeDirPath == "" {
+		// unmarshall to a NodeConfigControllerManager
+		configControllerManager := NodeConfigControllerManager{}
+		if err := json.Unmarshal(nodeConfig, &configControllerManager); err != nil {
+			logger.Warningf("unable to parse node config from controller manager: %+v", err)
+			return getDefaultFlexvolumeDir()
+		}
+		flexvolumeDirPath = configControllerManager.ComponentConfig.VolumeConfiguration.FlexVolumePluginDir
+		if flexvolumeDirPath == "" {
+			flexvolumeDirPath = getDefaultFlexvolumeDir()
+		}
+	}
+
+	logger.Infof("using flexvolume dir path: %s", flexvolumeDirPath)
+	return flexvolumeDirPath
+}
+
+func getDefaultFlexvolumeDir() string {
+	logger.Info("getting flexvolume dir path from provided env var")
+	flexvolumeDirPath := os.Getenv("FLEXVOLUME_DIR_PATH")
+	if flexvolumeDirPath == "" {
+		logger.Infof("flexvolume dir path is not provided. Defaulting to: %s", flexvolumeDefaultDirPath)
+		flexvolumeDirPath = flexvolumeDefaultDirPath
+	}
+	return flexvolumeDirPath
+}
+
+func getContainerImage(clientset kubernetes.Interface) (string, error) {
+
+	podName := os.Getenv(k8sutil.PodNameEnvVar)
```

Comment:
this is getting the container image that the operator is using, so that it can also be used in the agent's spec, right? maybe a simple comment to clear that up could help.  or maybe change the variable name to `operatorPodName` etc.

Reply:
Correct. 
i wanted to leave this function generic so that it can be reused if needed. Not just operator specific (although this is how it is being used now). 
I will add more comments